ambry-store/src/main/java/com.github.ambry.store/Log.java

```diff
@@ -13,161 +13,380 @@
   }
 
-  StoreMessageReadSet getView(List<BlobReadOptions> readOptions)
-      throws IOException {
-    return new StoreMessageReadSet(file, fileChannel, readOptions, currentWriteOffset.get());
+  /**
+   * {@inheritDoc}
+   * <p/>
+   * Appends the given {@code buffer} to the active log segment. The {@code buffer} will be written to a single log
+   * segment i.e. its data will not exist across segments.
+   * @param buffer The buffer from which data needs to be written from
+   * @return the number of bytes written.
+   * @throws IllegalArgumentException if the {@code buffer.remaining()} is greater than a single segment's size.
+   * @throws IllegalStateException if there no more capacity in the log.
+   * @throws IOException if there was an I/O error while writing.
+   */
+  @Override
+  public int appendFrom(ByteBuffer buffer) throws IOException {
+    rollOverIfRequired(buffer.remaining());
+    return activeSegment.appendFrom(buffer);
   }
 
-  public long sizeInBytes()
-      throws IOException {
-    return fileChannel.size();
+  /**
+   * {@inheritDoc}
+   * <p/>
+   * Appends the given data to the active log segment. The data will be written to a single log segment i.e. the data
+   * will not exist across segments.
+   * @param channel The channel from which data needs to be written from
+   * @param size The amount of data in bytes to be written from the channel
+   * @throws IllegalArgumentException if the {@code size} is greater than a single segment's size.
+   * @throws IllegalStateException if there no more capacity in the log.
+   * @throws IOException if there was an I/O error while writing.
+   */
+  @Override
+  public void appendFrom(ReadableByteChannel channel, long size) throws IOException {
+    rollOverIfRequired(size);
+    activeSegment.appendFrom(channel, size);
   }
 
-  public void setLogEndOffset(long endOffset)
-      throws IOException {
-    long fileSize = fileChannel.size();
-    if (endOffset < 0 || endOffset > fileSize) {
-      throw new IllegalArgumentException("Log : " + file.getAbsolutePath() + " endOffset " + endOffset +
-          " outside the file size " + fileSize);
+  /**
+   * Sets the active segment in the log.
+   * </p>
+   * Frees all segments that follow the active segment. Therefore, this should be
+   * used only after the active segment is conclusively determined.
+   * @param name the name of the log segment that is to be marked active.
+   * @throws IllegalArgumentException if there no segment with name {@code name}.
+   * @throws IOException if there is any I/O error freeing segments.
+   */
+  void setActiveSegment(String name) throws IOException {
+    if (!segmentsByName.containsKey(name)) {
+      throw new IllegalArgumentException("There is no log segment with name: " + name);
+    }
+    Map.Entry<String, LogSegment> higherEntry = segmentsByName.higherEntry(name);
+    while (higherEntry != null) {
+      free(higherEntry.getValue());
```

Comment:
Done. The file is currently deleted (impl of `free()` right now) but once the `DiskManager` implements pooling, the file will be returned to the pool.